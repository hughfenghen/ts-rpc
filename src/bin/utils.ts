import { ClassDeclaration, EnumDeclaration, ImportTypeNode, InterfaceDeclaration, ModuleDeclaration, Project, SourceFile, TypeAliasDeclaration, TypeReferenceNode, Node, ImportSpecifier, SyntaxKind, TypeNode } from 'ts-morph'

/**
 * 向SourceFile中插入Namespace，隔离命名空间，避免命名冲突
 */
export function insertNS (sf: SourceFile, appId: string): {
  ns: ModuleDeclaration
  rpcServices: InterfaceDeclaration
} {
  const appNS = sf.addModule({ name: `${appId}NS` })
  const appInterColl = appNS.addInterface({ name: 'App' })
  appNS.setIsExported(true)
  appInterColl.setIsExported(true)

  return {
    ns: appNS,
    rpcServices: appInterColl
  }
}

// 收集依赖树只考虑这四种场景
export type ITCDeclaration =
  | InterfaceDeclaration
  | TypeAliasDeclaration
  | ClassDeclaration
  | EnumDeclaration

export function collectTypeDeps (nodes: Node[], prj: Project): ITCDeclaration[] {
  const depsMap = new Set<ITCDeclaration>()

  nodes.forEach((node) => {
    if (node instanceof TypeReferenceNode) {
      findITDeclaration(node)
    } else if (isITCDeclaration(node)) {
      addDep(node)
    } else {
      // 深度优先遍历树，找到引用Type类型，然后找到 Declaration
      queryInTree(node)
    }
  })

  return Array.from(depsMap.values())

  function addDep (n: ITCDeclaration): void {
    // 避免重复，标准依赖无须添加
    if (depsMap.has(n) || isStandardType(n)) return
    const nodeName = n.getNameNode()?.getText()
    if (nodeName == null) throw new Error('dependency must be named')
    depsMap.add(n)
    // 被添加的依赖项，递归检查其依赖项
    queryInTree(n)
  }

  function queryInTree (n: Node): void {
    n.forEachChild(c => {
      if (c instanceof TypeReferenceNode) findITDeclaration(c)
      if (c instanceof ImportTypeNode) findIT4Import(c)

      queryInTree(c)
    })
  }

  // 收集依赖树只考虑这4种场景
  function isITCDeclaration (n: Node): n is ITCDeclaration {
    return n instanceof InterfaceDeclaration ||
      n instanceof TypeAliasDeclaration ||
      n instanceof ClassDeclaration ||
      n instanceof EnumDeclaration
  }

  function findITDeclaration (n: Node): void {
    const childrenIdf = n.getChildrenOfKind(SyntaxKind.Identifier)
    // 获取泛型的 Identifier
    const typeArgsIdf = deepFindTypeArgs(n)
      .map(t => t.getChildrenOfKind(SyntaxKind.Identifier))
      .flat()

    ; [...childrenIdf, ...typeArgsIdf].map(i => i.getSymbol())
      .map(s => s?.getDeclarations())
      .flat()
      .forEach(d => {
        if (d == null) return
        if (isITCDeclaration(d)) addDep(d)
        else if (d instanceof ImportSpecifier) findIT4ImportSpecifier(d)
      })
  }

  // 解析动态 import 函数
  function findIT4Import (n: ImportTypeNode): void {
    const fPath = n.getArgument().getText().slice(1, -1)
    const sf = prj.getSourceFile(sf => sf.getFilePath().includes(fPath))
    if (sf == null) throw new Error(`Could not find file ${fPath}`)
    const impName = n.getQualifier()?.getText() ?? ''
    const declaration = sf.getInterface(impName) ?? sf.getTypeAlias(impName) ?? sf.getClass(impName)
    if (declaration == null) throw Error(`Could not find interface, class or type (${impName}) in ${fPath}`)

    addDep(declaration)
  }

  // 解析import语法，从其他文件中查找依赖项
  function findIT4ImportSpecifier (is: ImportSpecifier): void {
    const impSf = is.getImportDeclaration().getModuleSpecifierSourceFile()
    if (impSf == null) throw new Error(`Could not find import var ${is.getText()}`)

    const impName = is.getText()
    const declaration = impSf.getInterface(impName) ?? impSf.getTypeAlias(impName) ?? impSf.getClass(impName)
    if (declaration == null) throw Error(`Could not find interface, class or type (${impName}) in ${impSf.getFilePath()}`)

    addDep(declaration)
  }
}

function deepFindTypeArgs (n: Node): TypeNode[] {
  if (n instanceof TypeReferenceNode) {
    const args = n.getTypeArguments()
    return args.concat(args.map(a => deepFindTypeArgs(a)).flat())
  }
  return []
}

export function isStandardType (n: Node): boolean {
  return /typescript\/lib|@types\/node/.test(n.getSourceFile().getFilePath())
}

export function addNode (container: ModuleDeclaration, n: Node): ITCDeclaration | null {
  if (n instanceof InterfaceDeclaration) {
    return container.addInterface(n.getStructure())
  } else if (n instanceof TypeAliasDeclaration) {
    return container.addTypeAlias(n.getStructure())
  } else if (n instanceof EnumDeclaration) {
    return container.addEnum(n.getStructure())
  } else if (n instanceof ClassDeclaration) {
    return container.addClass(n.getStructure())
  }
  return null
}
