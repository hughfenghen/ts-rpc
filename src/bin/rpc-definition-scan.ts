import glob from 'glob'
import { Project, Node, ClassDeclaration, FunctionDeclaration, MethodDeclaration, MethodSignature } from 'ts-morph'
import path from 'path'
import { existsSync } from 'fs'
import { IScanResult, TRPCMetaData } from '../common'
import { addNode, collectTypeDeps } from './utils'

export function scan (
  filePaths: string[],
  appId: string,
  opts?: {
    tsConfigFilePath?: string
  }
): IScanResult {
  const files = filePaths.map(f => glob.sync(f)).flat()
  const prj = new Project({
    tsConfigFilePath: opts?.tsConfigFilePath
  })
  files.forEach(file => {
    return prj.addSourceFileAtPath(file)
  })

  // 单测阶段
  const indexTs = path.resolve(__dirname, '../server/index.ts')
  // 编译后
  const indexDTs = path.resolve(__dirname, '../server/index.d.ts')
  const serverSf = prj.addSourceFileAtPath(
    existsSync(indexDTs) ? indexDTs : indexTs
  )
  const rpcServiceDef = serverSf.getFunction('RPCService')
  const rpcMethodDef = serverSf.getFunction('RPCMethod')
  if (rpcServiceDef == null || rpcMethodDef == null) throw Error('Cannot find RPCService or RPCMethod definition')

  const refedClasses = rpcServiceDef?.findReferencesAsNodes()
    // symbol <- callexpress <- decorator <- class
    .map(ref => ref.getParent()?.getParent()?.getParent())
    .filter(n => Node.isClassDeclaration(n)) as ClassDeclaration[]

  const genSf = prj.createSourceFile(path.resolve(__dirname, '_protocol-file-memory_.ts'), {
  }, { overwrite: true })

  // namespace 避免命名冲突
  const appNS = genSf.addModule({ name: `${appId}NS` })
  const appInterColl = appNS.addInterface({ name: 'App' })
  appNS.setIsExported(true)
  appInterColl.setIsExported(true)
  // 导出应用下聚合的 Service： export type AppID = AppIdNs.App;
  genSf.addTypeAlias({
    name: appId,
    type: `${appId}NS.App`,
    isExported: true
  })
  // 收集接口返回类型，用于 client 生成 json-schema
  const retTypes = appNS.addInterface({ name: 'APIReturnTypes' })
  retTypes.setIsExported(true)

  const rpcMetaData: TRPCMetaData = []
  // 已添加到 appNS 中的依赖，避免依赖项重复
  const addedDepIds: string[] = []
  // 遍历 class，将 class 转换为 interface，并将其依赖添加到 appNS
  refedClasses.forEach((c) => {
    const className = c.getName()

    if (className == null) throw Error('RPCService must be applied to a named class')
    if (appNS.getInterface(className) != null) throw Error(`RPCService marks duplicate class names: ${className}`)

    const methods = findRPCMethods(c, rpcMethodDef)
    if (methods.length === 0) {
      throw Error(`RPCService(${className}) must have at least one method`)
    }
    rpcMetaData.push({
      name: className,
      path: c.getSourceFile().getFilePath(),
      methods: methods.map(m => ({
        name: m.getName(),
        decorators: m.getDecorators().map(d => d.getText())
      }))
    })

    // 将 class 转换为 interface，模拟 rpc 的 protocol 声明
    const inter = appNS.addInterface({ name: className })
    inter.addJsDocs(c.getJsDocs().map(doc => doc.getStructure()))

    methods.forEach((m) => {
      // 移除原有decorator, 避免 bug：https://github.com/dsherret/ts-morph/issues/1214
      m.getDecorators().forEach(d => d.remove())
      // 移除 paramster 的 decorator，简化接口信息
      m.getParameters().forEach(p => {
        p.getDecorators().forEach(d => d.remove())
        // 移除参数初始值，class 将被转换为 interface， 而interface不支持初始值
        p.removeInitializer()
      })

      const ms = m.getSignature().getDeclaration() as MethodSignature
      const addedM = inter.addMethod(ms.getStructure())
      let rtText = addedM.getReturnTypeNode()?.getText()
      if (rtText == null) throw new Error(`Could not find method (${m.getName()}) return type`)

      retTypes.addProperty({ name: `'${className}.${m.getName()}'`, type: rtText })

      // 远程调用，返回值都是 Promise
      if (!/^Promise<.+>$/.test(rtText)) {
        rtText = `Promise<${rtText}>`
      }
      addedM.setReturnType(rtText)
    })

    collectTypeDeps(
      methods.map(m => [...m.getParameters(), m.getReturnTypeNodeOrThrow()]).flat(),
      prj
    ).forEach(it => {
      const nodeName = it.getNameNode()?.getText()
      if (nodeName == null) throw new Error('dependency must be named')

      const sid = it.getSourceFile().getFilePath() + '/' + nodeName
      // 避免重复
      if (addedDepIds.includes(sid)) return

      if (addedDepIds.some(sid => sid.endsWith(`/${nodeName}`))) {
        console.warn(`Named duplicate: ${nodeName}`)
        return
      }
      addedDepIds.push(sid)

      // 添加 method 依赖的类型
      const added = addNode(appNS, it)
      if (added instanceof ClassDeclaration) {
        // 只保留 class 的方法, 移除属性、class 的decorator 信息
        added.getDecorators().forEach(d => d.remove())
        added.getProperties()
          .forEach(p => p.getDecorators().forEach(d => d.remove()))
      }
      if (added == null) {
        console.warn(`unknown deps type: ${it.getText()}`)
      }
      added?.setIsExported(true)
    })

    appInterColl.addProperty({ name: className, type: className })
  })

  return {
    dts: genSf.getFullText(),
    meta: rpcMetaData
  }
}

function findRPCMethods (service: ClassDeclaration, rpcMethodDef: FunctionDeclaration): MethodDeclaration[] {
  return service.getMethods()
    .filter(m => m.getDecorators().some(
      d => {
        try {
          return d.getNameNode()
            .getDefinitionNodes()
            .some(n => n === rpcMethodDef)
        } catch (e) {
          return false
        }
      }
    ))
}
