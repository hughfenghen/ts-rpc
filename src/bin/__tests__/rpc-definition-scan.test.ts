import path from 'path'
import { Project } from 'ts-morph'
import { scan } from '../rpc-definition-scan'
import { collectTypeDeps } from '../utils'

console.warn = jest.fn()

test('scan', () => {
  const { dts, meta } = scan([path.resolve(__dirname, '*.ts')], 'ScanTest', {
    tsConfigFilePath: path.resolve(__dirname, 'tsconfig.json')
  })

  expect(dts).toMatchSnapshot()
  expect(meta).toEqual([{
    name: 'User',
    path: expect.stringContaining('/src/bin/__tests__/controller-example.ts'),
    methods: [
      { name: 'getInfoById1', decorators: ['@RPCMethod()'] },
      { name: 'getInfoById2', decorators: ['@RPCMethod()'] },
      { name: 'getInfoById3', decorators: ['@RPCMethod()'] },
      { name: 'getUnreadMsg', decorators: ['@(OtherDecorator())', '@RPCMethod()'] }
    ]
  }, {
    name: 'Foo',
    path: expect.stringContaining('/src/bin/__tests__/controller-example.ts'),
    methods: [{ name: 'bar', decorators: ['@RPCMethod()'] }]
  }, {
    name: 'User1',
    path: expect.stringContaining('/src/bin/__tests__/controller-example1.ts'),
    methods: [
      { name: 'getInfoById1', decorators: ['@RPCMethod()'] }
    ]
  }])
})

test('collectTypeDeps', () => {
  const prj = new Project({ compilerOptions: { declaration: true } })
  const sf = prj.createSourceFile('test.ts', `
    interface A {
      b: B
    }
    interface B {
      a: A
    }

    interface X {
      a: A
      b: B
      ab: A | B
      s: string
      n: number
      f: () => any
    }
    type Y = A | B | null | undefined
  `)
  const iDeps = collectTypeDeps(sf.getInterfaceOrThrow('X'), prj)
  expect(iDeps.map(i => i.getName())).toEqual(['X', 'A', 'B'])

  const tDeps = collectTypeDeps(sf.getTypeAliasOrThrow('Y'), prj)
  expect(tDeps.map(t => t.getName())).toEqual(['Y', 'A', 'B'])
})
