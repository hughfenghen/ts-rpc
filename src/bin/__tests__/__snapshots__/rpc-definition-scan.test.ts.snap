// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scan 1`] = `
"export namespace ScanTestNS {
    type UnwrapPromise<T> = T extends Promise<infer U> ? U : T;

    export interface App {
        User: User;
        User1: User1;
    }

    export interface APIReturnTypes {
        'User.getInfoById1': UnwrapPromise<CUserInfo & IDuplicate>;
        'User.getInfoById2': UnwrapPromise<IUserInfo & { gender: Gender1 }>;
        'User.getInfoById3': UnwrapPromise<TUserInfo>;
        'User.getUnreadMsg': UnwrapPromise<Promise<string[]>>;
        'User1.getInfoById1': UnwrapPromise<CUserInfo & IDuplicate>;
    }

    /**
     * service doc
     */
    interface User {
        /**
         * method doc
         */
        getInfoById1(id: string): CUserInfo & IDuplicate;
        getInfoById2(id: string): IUserInfo & { gender: Gender1 };
        getInfoById3(id: string): TUserInfo;
        getUnreadMsg(id: string): Promise<string[]>;
    }

    /**
     * service doc
     * 验证公共依赖 CUserInfo
     */
    interface User1 {
        /**
         * method doc
         */
        getInfoById1(id: string): CUserInfo & IDuplicate;
    }

    export class CUserInfo {
        name: string;
        age: Date;
        avatar: string;
    }

    export type Gender = Gender1.Male | Gender1.Female;
    export type TUserInfo = CUserInfo & IUserInfo;

    /**
     * 与其他文件中的变量命名冲突，只保留一份，扫描时提供警告
     */
    export interface IDuplicate {
        nothing: string;
    }

    export interface IUserInfo {
        name: string;
        age: Date;
        avatar: string;
        gender: Gender;
    }

    /**
     * 注释：性别
     */
    export enum Gender1 {
        /**
         * 枚举注释
         */
        Male = 'male',
        Female = 'female'
    }
}

export type ScanTest = ScanTestNS.App;
"
`;

exports[`scan 2`] = `
"[
  {
    \\"name\\": \\"User\\",
    \\"path\\": \\"/Users/fenghen/bili-space/ts-rpc/src/bin/__tests__/controller-example.ts\\",
    \\"methods\\": [
      {
        \\"name\\": \\"getInfoById1\\",
        \\"decorators\\": [
          \\"@RPCMethod()\\"
        ],
        \\"retSchema\\": {
          \\"allOf\\": [
            {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                },
                \\"age\\": {
                  \\"type\\": \\"string\\",
                  \\"format\\": \\"date-time\\"
                },
                \\"avatar\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"age\\",
                \\"avatar\\",
                \\"name\\"
              ]
            },
            {
              \\"description\\": \\"与其他文件中的变量命名冲突，只保留一份，扫描时提供警告\\",
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"nothing\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"nothing\\"
              ]
            }
          ]
        }
      },
      {
        \\"name\\": \\"getInfoById2\\",
        \\"decorators\\": [
          \\"@RPCMethod()\\"
        ],
        \\"retSchema\\": {
          \\"allOf\\": [
            {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                },
                \\"age\\": {
                  \\"type\\": \\"string\\",
                  \\"format\\": \\"date-time\\"
                },
                \\"avatar\\": {
                  \\"type\\": \\"string\\"
                },
                \\"gender\\": {
                  \\"enum\\": [
                    \\"female\\",
                    \\"male\\"
                  ],
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"age\\",
                \\"avatar\\",
                \\"gender\\",
                \\"name\\"
              ]
            },
            {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"gender\\": {
                  \\"description\\": \\"注释：性别\\",
                  \\"enum\\": [
                    \\"female\\",
                    \\"male\\"
                  ],
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"gender\\"
              ]
            }
          ]
        }
      },
      {
        \\"name\\": \\"getInfoById3\\",
        \\"decorators\\": [
          \\"@RPCMethod()\\"
        ],
        \\"retSchema\\": {
          \\"allOf\\": [
            {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                },
                \\"age\\": {
                  \\"type\\": \\"string\\",
                  \\"format\\": \\"date-time\\"
                },
                \\"avatar\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"age\\",
                \\"avatar\\",
                \\"name\\"
              ]
            },
            {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                },
                \\"age\\": {
                  \\"type\\": \\"string\\",
                  \\"format\\": \\"date-time\\"
                },
                \\"avatar\\": {
                  \\"type\\": \\"string\\"
                },
                \\"gender\\": {
                  \\"enum\\": [
                    \\"female\\",
                    \\"male\\"
                  ],
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"age\\",
                \\"avatar\\",
                \\"gender\\",
                \\"name\\"
              ]
            }
          ]
        }
      },
      {
        \\"name\\": \\"getUnreadMsg\\",
        \\"decorators\\": [
          \\"@(OtherDecorator())\\",
          \\"@RPCMethod()\\"
        ],
        \\"retSchema\\": {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      }
    ]
  },
  {
    \\"name\\": \\"User1\\",
    \\"path\\": \\"/Users/fenghen/bili-space/ts-rpc/src/bin/__tests__/controller-example1.ts\\",
    \\"methods\\": [
      {
        \\"name\\": \\"getInfoById1\\",
        \\"decorators\\": [
          \\"@RPCMethod()\\"
        ],
        \\"retSchema\\": {
          \\"allOf\\": [
            {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                },
                \\"age\\": {
                  \\"type\\": \\"string\\",
                  \\"format\\": \\"date-time\\"
                },
                \\"avatar\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"age\\",
                \\"avatar\\",
                \\"name\\"
              ]
            },
            {
              \\"description\\": \\"与其他文件中的变量命名冲突，只保留一份，扫描时提供警告\\",
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"nothing\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"nothing\\"
              ]
            }
          ]
        }
      }
    ]
  }
]"
`;
